File: quantsolve-server/src/engine/parser.js
// engine/parser.js

const { TokenType } = require("./lexer");
const {
    NumberNode,
    VariableNode,
    BinaryOpNode,
    UnaryOpNode,
} = require("./ast");
const { EngineError, ErrorCode } = require("./errors");

function containsVariable(node) {
    if (!node) return false;
    if (node.type === "Variable") return true;
    if (node.type === "Number") return false;
    if (node.type === "UnaryOp") return containsVariable(node.operand);
    if (node.type === "BinaryOp") {
        return containsVariable(node.left) || containsVariable(node.right);
    }
    return false;
}

function evalConstant(node) {
    if (!node) return null;
    if (node.type === "Number") return node.value;
    if (node.type === "UnaryOp" && node.op === "-") {
        const v = evalConstant(node.operand);
        return v === null ? null : -v;
    }
    if (node.type === "BinaryOp") {
        const l = evalConstant(node.left);
        const r = evalConstant(node.right);
        if (l === null || r === null) return null;
        if (node.operator === "+") return l + r;
        if (node.operator === "-") return l - r;
        if (node.operator === "*") return l * r;
        if (node.operator === "/") return r === 0 ? null : l / r;
        if (node.operator === "^") {
            if (!Number.isInteger(r) || r < 0) return null;
            return l ** r;
        }
    }
    return null;
}

class Parser {
    constructor(tokens) {
        this.tokens = tokens;
        this.pos = 0;
    }

    peek() {
        return (
            this.tokens[this.pos] || {
                type: TokenType.EOF,
                value: "EOF",
                position: -1,
            }
        );
    }

    eat(expectedType) {
        const tok = this.peek();
        if (tok.type !== expectedType) {
            throw new EngineError(
                ErrorCode.UNEXPECTED_TOKEN,
                tok.value ?? tok.type,
                tok.position,
            );
        }
        this.pos += 1;
        return tok;
    }

    // atom = NUMBER | VARIABLE | "(" expression ")" | "-" atom
    parseFactor() {
        const tok = this.peek();

        if (tok.type === TokenType.MINUS) {
            this.eat(TokenType.MINUS);
            const operand = this.parseFactor();
            return new UnaryOpNode("-", operand);
        }

        if (tok.type === TokenType.NUMBER) {
            this.eat(TokenType.NUMBER);
            return new NumberNode(tok.value);
        }

        if (tok.type === TokenType.VARIABLE) {
            this.eat(TokenType.VARIABLE);
            return new VariableNode(tok.value);
        }

        if (tok.type === TokenType.LPAREN) {
            this.eat(TokenType.LPAREN);
            const inner = this.parseExpression();
            this.eat(TokenType.RPAREN);
            return inner;
        }

        throw new EngineError(
            ErrorCode.UNEXPECTED_TOKEN,
            tok.value ?? tok.type,
            tok.position,
        );
    }

    // power = factor ("^" factor)*
    parsePower() {
        let node = this.parseFactor();

        while (this.peek().type === TokenType.POW) {
            this.eat(TokenType.POW);
            node = new BinaryOpNode(node, "^", this.parseFactor());
        }

        return node;
    }

    // term = power (("*" | "/") power)*
    parseTerm() {
        let node = this.parsePower();

        while (
            this.peek().type === TokenType.MUL ||
            this.peek().type === TokenType.DIV
        ) {
            const opTok = this.peek();

            if (opTok.type === TokenType.MUL) {
                this.eat(TokenType.MUL);
                node = new BinaryOpNode(node, "*", this.parsePower());
            } else {
                this.eat(TokenType.DIV);
                const denomNode = this.parsePower();

                if (containsVariable(denomNode)) {
                    const varName =
                        denomNode.type === "Variable"
                            ? denomNode.name
                            : "unknown";
                    throw new EngineError(
                        ErrorCode.VARIABLE_IN_DENOMINATOR,
                        varName,
                    );
                }

                const denomValue = evalConstant(denomNode);
                if (denomValue === 0) {
                    throw new EngineError(ErrorCode.DIVISION_BY_ZERO);
                }

                node = new BinaryOpNode(node, "/", denomNode);
            }
        }

        return node;
    }

    // expression = term (("+" | "-") term)*
    parseExpression() {
        let node = this.parseTerm();

        while (
            this.peek().type === TokenType.PLUS ||
            this.peek().type === TokenType.MINUS
        ) {
            const opTok = this.peek();
            if (opTok.type === TokenType.PLUS) {
                this.eat(TokenType.PLUS);
                node = new BinaryOpNode(node, "+", this.parseTerm());
            } else {
                this.eat(TokenType.MINUS);
                node = new BinaryOpNode(node, "-", this.parseTerm());
            }
        }

        return node;
    }

    // equation = expression "=" expression EOF
    parseEquation() {
        const left = this.parseExpression();
        this.eat(TokenType.EQUAL);
        const right = this.parseExpression();

        if (this.peek().type !== TokenType.EOF) {
            throw new EngineError(ErrorCode.TRAILING_TOKENS);
        }

        return { left, right };
    }
}

module.exports = Parser;

File: quantsolve-server/src/engine-cpp/src/parser.cpp
#include "quantsolve/parser.hpp"

#include "quantsolve/errors.hpp"

#include <cmath>
#include <optional>

namespace quantsolve {

namespace {

std::optional<double> eval_constant_or_null(const ASTPtr& node) {
    if (!node) return std::nullopt;

    if (node->type == NodeType::Number) return static_cast<double>(node->value);

    if (node->type == NodeType::Variable) return std::nullopt;

    if (node->type == NodeType::UnaryOp && node->op == '-') {
        auto v = eval_constant_or_null(node->operand);
        if (!v.has_value()) return std::nullopt;
        return -(*v);
    }

    if (node->type == NodeType::BinaryOp) {
        auto l = eval_constant_or_null(node->left);
        auto r = eval_constant_or_null(node->right);
        if (!l.has_value() || !r.has_value()) return std::nullopt;

        switch (node->op) {
        case '+': return *l + *r;
        case '-': return *l - *r;
        case '*': return *l * *r;
        case '/':
            if (*r == 0.0) return std::nullopt;
            return *l / *r;
        case '^':
            if (*r < 0.0 || std::llround(*r) != *r) return std::nullopt;
            return std::pow(*l, *r);
        default: return std::nullopt;
        }
    }

    return std::nullopt;
}

} // namespace

bool contains_variable(const ASTPtr& node) {
    if (!node) return false;
    if (node->type == NodeType::Variable) return true;
    if (node->type == NodeType::Number) return false;
    if (node->type == NodeType::UnaryOp) return contains_variable(node->operand);
    if (node->type == NodeType::BinaryOp) {
        return contains_variable(node->left) || contains_variable(node->right);
    }
    return false;
}

Parser::Parser(std::vector<Token> tokens)
    : tokens_(std::move(tokens)) {}

const Token& Parser::peek() const {
    static const Token eof{TokenType::EOF_TOKEN, "EOF", -1};
    if (pos_ >= tokens_.size()) return eof;
    return tokens_[pos_];
}

Token Parser::eat(TokenType expected) {
    const Token tok = peek();
    if (tok.type != expected) {
        throw EngineError(ErrorCode::UNEXPECTED_TOKEN, {tok.value, std::to_string(tok.position)});
    }
    ++pos_;
    return tok;
}

ASTPtr Parser::parse_factor() {
    const Token tok = peek();

    if (tok.type == TokenType::MINUS) {
        eat(TokenType::MINUS);
        return make_unary('-', parse_factor());
    }

    if (tok.type == TokenType::NUMBER) {
        eat(TokenType::NUMBER);
        return make_number(std::stoll(tok.value));
    }

    if (tok.type == TokenType::VARIABLE) {
        eat(TokenType::VARIABLE);
        return make_variable(tok.value);
    }

    if (tok.type == TokenType::LPAREN) {
        eat(TokenType::LPAREN);
        auto inner = parse_expression();
        eat(TokenType::RPAREN);
        return inner;
    }

    throw EngineError(ErrorCode::UNEXPECTED_TOKEN, {tok.value, std::to_string(tok.position)});
}

ASTPtr Parser::parse_power() {
    ASTPtr node = parse_factor();

    while (peek().type == TokenType::POW) {
        eat(TokenType::POW);
        node = make_binary(node, '^', parse_factor());
    }

    return node;
}

ASTPtr Parser::parse_term() {
    ASTPtr node = parse_power();

    while (peek().type == TokenType::MUL || peek().type == TokenType::DIV) {
        Token op = peek();

        if (op.type == TokenType::MUL) {
            eat(TokenType::MUL);
            node = make_binary(node, '*', parse_power());
        } else {
            eat(TokenType::DIV);
            ASTPtr denom = parse_power();

            if (contains_variable(denom)) {
                std::string var_name = "unknown";
                if (denom && denom->type == NodeType::Variable) {
                    var_name = denom->name;
                }
                throw EngineError(ErrorCode::VARIABLE_IN_DENOMINATOR, {var_name});
            }

            auto denom_val = eval_constant_or_null(denom);
            if (denom_val.has_value() && *denom_val == 0.0) {
                throw EngineError(ErrorCode::DIVISION_BY_ZERO);
            }

            node = make_binary(node, '/', denom);
        }
    }

    return node;
}

ASTPtr Parser::parse_expression() {
    ASTPtr node = parse_term();

    while (peek().type == TokenType::PLUS || peek().type == TokenType::MINUS) {
        Token op = peek();
        if (op.type == TokenType::PLUS) {
            eat(TokenType::PLUS);
            node = make_binary(node, '+', parse_term());
        } else {
            eat(TokenType::MINUS);
            node = make_binary(node, '-', parse_term());
        }
    }

    return node;
}

std::pair<ASTPtr, ASTPtr> Parser::parse_equation() {
    ASTPtr left = parse_expression();
    eat(TokenType::EQUAL);
    ASTPtr right = parse_expression();

    if (peek().type != TokenType::EOF_TOKEN) {
        throw EngineError(ErrorCode::TRAILING_TOKENS);
    }

    return {left, right};
}

} // namespace quantsolve
